network: Avoid creating link files for 'nbft' interfaces

tbzatek opened the pull request:
The dracut networking and the 95nvmf module are using ifname= kernel commandline arguments to pass desired network interface setup as parsed from the ACPI NBFT table. These are highly volatile values and often changing upon each boot.

Thus avoid writing /etc/systemd/network interface link files by Anaconda for any 'nbft*' network interface as these are getting packed in initramfs, creating havoc and conflicts with the actual networking setup during early boot.

----

This is a clone of #5733.

jkonecny12 commented:
/kickstart-test --testtype smoke

jkonecny12 commented:
/kickstart-test --testtype network

jkonecny12 reviewed (approved):
LGTM. I'm setting @rvykydal for review just in case he we will show up.

rvykydal reviewed (commented):
I have 3 comments:

- Ignoring at this level instead of higher, perhaps here: https://github.com/rhinstaller/anaconda/blob/d27cd3efedade84f999627ef423f5b4b5b243e9e/pyanaconda/modules/network/network.py#L762 the ifname option will be used when applying kickstart configuration of nbft device or creating default connection profile for it. Which means the connection/profile will be bound to the mac address. Not sure if it is a problem, we can also address it when / if it occurs.
- ~I'd add a debug log message about ignoring the device~
- I'd add a test for this case (I'll come with up with some: https://github.com/rhinstaller/anaconda/pull/5771)

rvykydal commented on pyanaconda/modules/network/installation.py:
```diff
@@ -186,6 +186,9 @@ def _write_interface_rename_config(self, root, ifname_option_values, overwrite):
         for ifname_value in ifname_option_values:
             iface, mac = ifname_value.split(":", 1)
+            # Avoid link files for 'nbft' interfaces as it breaks Boot from NVM over TCP
+            if iface.startswith('nbft'):
```
Maybe we should not create the 'target_system_dir' few lines above if the list contains only "nbft*" values, so perhaps doing 
```
ifname_option_values = [value in ifname_option_values if not ifname_option_value.startswith('nbft')]
```
at the beginning of the method.

tbzatek commented:
Tested `RHEL-9.5.0-20240731.4` with the original (simple) patch applied through `inst.updates`. Attaching logs grabbed from installer environment right before reboot: [logs-inst-env.tar.gz](https://github.com/user-attachments/files/16470867/logs-inst-env.tar.gz)

Post-reboot in the installed system:
- nothing network-specific in `/root/anaconda-ks.cfg`
- no `/etc/systemd/network/10-anaconda-ifname-*` files (the patch works)
- NM system connection files got created though:
```
# cat /etc/NetworkManager/system-connections/nbft0.nmconnection 
[connection]
id=nbft0
uuid=aabe2131-b75f-464e-b71f-d04263a085b3
type=ethernet
autoconnect-priority=-100
autoconnect-retries=1
interface-name=nbft0
timestamp=1722601497

[ethernet]
mac-address=52:54:00:72:C5:AE

[ipv4]
address1=192.168.122.58/24,192.168.122.1
may-fail=false
method=manual

[ipv6]
addr-gen-mode=eui64
method=disabled

[proxy]

[user]
org.freedesktop.NetworkManager.origin=nm-initrd-generator
```

Fortunately looks like these files are ignored when network config is supplied by dracut. Tested changing IP addresses, this file still had an old one. So it looks persistent, and that's bad. Removed the files and everything worked as expected, both with static addresses as well with DHCP, reacting to link changes and renegotiating 

...that led me to: https://github.com/rhinstaller/anaconda/blob/60d39b18399ce0ecf8ac187ae5d88a84d4a3750c/pyanaconda/modules/network/device_configuration.py#L162-L187

and 

https://github.com/rhinstaller/anaconda/blob/60d39b18399ce0ecf8ac187ae5d88a84d4a3750c/pyanaconda/modules/network/device_configuration.py#L308-L336

In theory, anything iBFT-specific mostly applies to nBFT as well (though I'm totally unaware how iBFT support looks like in Anaconda)

So I would vote for avoid creating these files. Not a RHEL 9.5 material though, needs to be tested thoroughly.

tbzatek commented:
I've experimenting with including `nbft` in all `ibft` tests but the result is essentially unchanged - the `.nmconnection` files are still being created.

```patch
diff --git a/pyanaconda/core/regexes.py b/pyanaconda/core/regexes.py
index cc00702e3a..bcd1c3ea6f 100644
--- a/pyanaconda/core/regexes.py
+++ b/pyanaconda/core/regexes.py
@@ -166,8 +166,8 @@ ISCSI_IQN_NAME_REGEX = re.compile(r'^iqn\.\d{4}-\d{2}((?<!-)\.(?!-)[a-zA-Z0-9\-]
 #    a. The format is "eui." followed by an EUI-64 identifier (16 ASCII-encoded hexadecimal digits).
 ISCSI_EUI_NAME_REGEX = re.compile(r'^eui\.[a-fA-F0-9]{16}$')
 
-# Device with this name was configured from ibft (and renamed) by dracut
-IBFT_CONFIGURED_DEVICE_NAME = re.compile(r'^ibft\d+$')
+# Device with this name was configured from ibft/nbft (and renamed) by dracut
+IBFT_CONFIGURED_DEVICE_NAME = re.compile(r'^[in]bft\d+$')
 
 # Regex to validate a DASD device number (full or partial).
 #
diff --git a/pyanaconda/modules/network/device_configuration.py b/pyanaconda/modules/network/device_configuration.py
index f97560b439..a349eca829 100644
--- a/pyanaconda/modules/network/device_configuration.py
+++ b/pyanaconda/modules/network/device_configuration.py
@@ -181,6 +181,9 @@ class DeviceConfigurations(object):
         elif device.get_iface().endswith(('-fcoe', '-fco', '-fc', '-f', '-')):
             decline_reason = "special FCoE vlan device"
 
+        elif device.get_iface().startswith('nbft'):
+            decline_reason = "nBFT device"
+
         # Ignore devices configured via iBFT, ie
         # devices with active read-only connections (created by NM for iBFT VLAN)
         elif self._has_read_only_active_connection(device):
diff --git a/pyanaconda/modules/network/installation.py b/pyanaconda/modules/network/installation.py
index 8202956028..4ac786c6e0 100644
--- a/pyanaconda/modules/network/installation.py
+++ b/pyanaconda/modules/network/installation.py
@@ -188,6 +188,9 @@ Name={}
 
         for ifname_value in ifname_option_values:
             iface, mac = ifname_value.split(":", 1)
+            # Avoid link files for 'nbft' interfaces as it breaks Boot from NVM over TCP
+            if iface.startswith('nbft'):
+                continue
             content = self.INTERFACE_RENAME_FILE_CONTENT_TEMPLATE.format(mac, iface)
             config_file = self.INTERFACE_RENAME_FILE_TEMPLATE.format(iface)
             config_file_path = util.join_paths(self.SYSTEMD_NETWORK_CONFIG_DIR, config_file)
diff --git a/pyanaconda/modules/network/network.py b/pyanaconda/modules/network/network.py
index a58fd9775a..6d9e400f72 100644
--- a/pyanaconda/modules/network/network.py
+++ b/pyanaconda/modules/network/network.py
@@ -680,7 +680,8 @@ class NetworkService(KickstartService):
         if 'BOOTIF' in kernel_args:
             self.bootif = kernel_args.get('BOOTIF')[3:].replace("-", ":").upper()
         if 'ifname' in kernel_args:
-            self.ifname_option_values = kernel_args.get("ifname").split()
+            self.ifname_option_values = [value for value in kernel_args.get("ifname").split()
+                                         if not value.startswith('nbft')]
         if 'noipv6' in kernel_args:
             self.disable_ipv6 = True
 
```

rvykydal commented:
Thank you for testing, you are brave looking into anaconda network config code, on the right way, we just need one more guard (maybe rather two to be safe). I added it to the https://github.com/rhinstaller/anaconda/pull/5771

To sum up what is there:
- Ignoring `ifname=nbftX` in device renaming (just updated a bit your original patch and added test)
- Ignoring nBFT devices in GUI - your proposal above a bit modified
- Ignoring nBFT devices when dumping network configuration (created in initramfs) - the  https://github.com/rhinstaller/anaconda/pull/5771/commits/3101fb12dfa674e78aa682c38a45aeda608a70cd
- Ignoring configuration of nBFT devices from kickstart - https://github.com/rhinstaller/anaconda/pull/5771/commits/f03d45ec1fb0676794239e075ff07e4800916fcd. We don't expect user to explicitly configure such devices with --device=nbftX (which would be considered unsupported) but the check is there rather to prevent some non-explicit configuration (`--device=link`) potentially resolved to an nBFT device.

With all that we don't need to ignore ifname=nbft completely in the module in _apply_boot_options (and I think module itself should not hide the data).
To by safe we could also add ignoring nbft configuration from kickstar

rvykydal commented:
Closing as obsoleted by https://github.com/rhinstaller/anaconda/pull/5771
